src/compiler/binder.ts

```diff
@@ -608,9 +609,21 @@ namespace ts {
         }
 
         function isNarrowableReference(expr: Expression): boolean {
-            return expr.kind === SyntaxKind.Identifier ||
-                expr.kind === SyntaxKind.ThisKeyword ||
-                expr.kind === SyntaxKind.PropertyAccessExpression && isNarrowableReference((<PropertyAccessExpression>expr).expression);
+            if (expr.kind === SyntaxKind.ElementAccessExpression) {
+                const argument = (expr as ElementAccessExpression).argumentExpression;
+                return argument.kind === SyntaxKind.StringLiteral || argument.kind === SyntaxKind.NumericLiteral || argument.kind === SyntaxKind.EnumMember;

```

Comment:
Actually, multiple levels of property access don't work either. So I think this function should actually be even simpler until we fix that bug.

```ts
interface Square {
  sub: { kind: "square; };
  size: number;
}
// etc ...
function area(s: Shape) {
    switch(s.sub.kind) {
        case "square": return s.size * s.size;
        case "rectangle": return s.width * s.height;
        case "circle": return Math.PI * s.radius * s.radius;
    }
}
```